tests/system/test_query_retry.py

```diff
@@ -0,0 +1,65 @@
+# Copyright (c) 2021 The PyBigQuery Authors
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy of
+# this software and associated documentation files (the "Software"), to deal in
+# the Software without restriction, including without limitation the rights to
+# use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
+# the Software, and to permit persons to whom the Software is furnished to do so,
+# subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+# FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+# COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+# IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+# CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+import contextlib
+import threading
+import time
+
+import google.cloud.bigquery
+
+
+def thread(func):
+    thread = threading.Thread(target=func, daemon=True)
+    thread.start()
+    return thread
+
+
+def test_query_retry_539(bigquery_client, dataset_id):
+    """
+    Test semantic retry
+
+    See: https://github.com/googleapis/python-bigquery/issues/539
+    """
+    from google.api_core import exceptions
+    from google.api_core.retry import if_exception_type, Retry
+
+    table_name = f"{dataset_id}.t539"
+    job = bigquery_client.query(f"select count(*) from {table_name}")
+    job_id = job.job_id
+
+    # We can already know that the job failed, but we're not supposed
+    # to find out until we call result, which is where retry happend
+    assert job.done()
+    assert job.exception() is not None
+
+    @thread
+    def create_table():
+        time.sleep(1)
+        with contextlib.closing(google.cloud.bigquery.Client()) as client:
+            client.query(f"create table {table_name} (id int64)")
+
+    retry_policy = Retry(predicate=if_exception_type(exceptions.NotFound))
+    [[count]] = list(job.result(retry=retry_policy))
```

Comment:
Oh, is the `retry` used to determine if we should re-issue a query? I think we should have a separate object for that, since retrying individual API requests can still be useful in cases where we don't want to retry the whole query if its failed.

Reply:
OK, I changed it to use `DEFAULT_RETRY` for the 2 sub-requests, however...

There's a problem here.  I don't think users understand retry.  They expect it to deal with transient semantic errors, like NotFound and rate limiting.  They don't get that network errors and application errors are handled separately.  They probably aren't aware of what network errors they should worry about.  It took me a while to grok the difference between errors raised by `api_request` and other errors that cause a job to be in a failed state.  I'm not 100% sure I understand the difference now. 0.5 ;)

IMO, it would **not** help to let them supply two different retries.

Some problems with this PR as it stands ATM:

1. The meaning of the `retry` argument to `result()` depends on whether a job id was supplied to `query`.  This is bad. :)
2. Users probably expect to specify application-level retry when calling `query`, not when calling `result`.

A fundamental problem, IMO, is that there are two kinds of retry and the distinction between them is subtle.

Maybe, a breaking change we could make, since we're making breaking changes :), would be to:

1. Define and use a new retry for network retry. The user doesn't influence this.
2. Use the `retry` argument to `query` and `result` to specify how failed jobs should be retried.
   - Default them to a new sensible default, perhaps limited to rate-limit errors.
   - The retry passed to query is only used as the default retry for `result`.
3. If a user supplies a `job id` to `query`, then error if they pass a `retry` argument to `query` or `result`, to the effect that their job isn't retryable, because the supplied a custom job id. 

I suspect we should brainstorm this a bit.

Reply:
I'd hate for us to retry the query if there was just a transient API error, so I do think we need separate logic.

> A fundamental problem, IMO, is that there are two kinds of retry and the distinction between them is subtle.

I agree. I'm hoping the with good defaults, most users won't have to worry about it. But I also don't think it's something we should hide from them. It's not intuitive and requires some detailed knowledge about how the API works to get right, but I think that's true for custom retry logic in general.

Reply:
> The retry passed to query is only used as the default retry for result.

This could work. Especially if we use the `jobs.query` API. https://github.com/googleapis/python-bigquery/issues/589 This is all way too complicated. 😭

Reply:
> I'd hate for us to retry the query if there was just a transient API error, so I do think we need separate logic.

I'm not suggesting we don't retry API/network errors. I was suggesting we don't give users control over them.

However, in the interest of backward compatibility, I propose:
- Document the existing `retry` argument as API retry. IN the documentation, I'd say something to the effect of "this isn't the retry you're looking for." 
- Add a new `job_retry` argument to `query` and `result` to retry failed jobs. (The `job_retry` passed to `query` is simply the default for `result`.)
- Error if `job_retry` is provided for non-retryable jobs, due to user-suplied job id.

Reply:
I like this proposal.

Reply:
Done, see https://github.com/googleapis/python-bigquery/pull/837/commits/4c6ef5bd48b51c4b676114a46d687158b30fb8c3.